### PR TITLE
Stop ignoring expected note/help messages in compiletest suite.

### PR DESCRIPTION
**Pull request by frewsxcv**
Original issue: https://github.com/rust-lang/rust/issues/21195

Relevant PR: https://github.com/rust-lang/rust/pull/30778

Prior to this commit, if a compiletest testcase included the text
"HELP:" or "NOTE:" (note the colons), then it would indicate to the
compiletest suite that we should verify "help" and "note" expected
messages.

This commit updates this check to also check "HELP" and "NOTE" (not the
absense of colons) so that we always verify "help" and "note" expected
messages.

### CODE REVIEW

File: src/compiletest/runtest.rs [hunk 1]
```diff
@@ -1013,8 +1013,8 @@ fn check_expected_errors(revision: Option<&str>,
         expected_errors.iter()
                         .fold((false, false),
                               |(acc_help, acc_note), ee|
-                                  (acc_help || ee.kind == "help:", acc_note ||
-                                   ee.kind == "note:"));
+                                  (acc_help || ee.kind == "help:" || ee.kind == "help",
+                                   acc_note || ee.kind == "note:" || ee.kind == "note"));
```
Comment (frewsxcv) [inline src/compiletest/runtest.rs]: I have a separate branch that cleans up this "kind" logic by using enums. I plan to open it up as a follow-up PR if this merges.

File: src/test/compile-fail/issue-25386.rs [hunk 2]
```diff
@@ -8,6 +8,10 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+// error-pattern: in this expansion of format_args!
+// error-pattern: in this expansion of print! (defined in <std macros>)
+// field `c_object` of struct `stuff::Item` is private
```
Comment (frewsxcv) [inline src/test/compile-fail/issue-25386.rs]: I'm pretty sure I did this incorrectly. A couple errors in this testcase have errors that report on `<std macros>` and I'm not sure the best way to handle it.

**Comment by frewsxcv** [comment]
A little background on this PR: I did a bit of refactoring on compiletest and realized that some help/note messages were being ignored. I noticed [this conditional](https://github.com/rust-lang/rust/blob/483fc715c6ea7cdd15d41d79df9def97401cd699/src/compiletest/runtest.rs#L1016-L1017) and assumed it was a mistake to not also check `"note"` and `"help"`, so I fixed the conditional and the relevant test cases (this PR). It wasn't until after I was done that I noticed [this issue](https://github.com/rust-lang/rust/issues/21195) where it makes it sound like it's opt-in, presumably because it's too much work to add the extra note/help messages? In my opinion, there's no harm in adding them and it could catch diagnostic errors in the future.

**Comment by frewsxcv** [comment]
On second thought, I think it is still opt-in, but my changes make the opt-in work for both `"NOTE"` and `"HELP"` in addition to `"NOTE:"` and `"HELP:"`

File: src/compiletest/runtest.rs (hunk 1 above)
Comment (fhahn) [inline src/compiletest/runtest.rs]: This should  make the opt-in work for `note:` and `help:` as well, but it is still opt-in and notes and help messages are only checked in files with `help:` and `note:` annotations.

When making this changes, it seems like I overlooked `note:` and `help:` (and assumed there is more normalization going on)

File: src/compiletest/runtest.rs [hunk 3]
```diff
@@ -1013,8 +1013,8 @@ fn check_expected_errors(revision: Option<&str>,
         expected_errors.iter()
                         .fold((false, false),
```
Comment (nikomatsakis) [inline src/compiletest/runtest.rs]: Pre-existing, but can't we use `any` here instead of `fold`? Seems like it more clearly expresses the intent.

Comment (frewsxcv) [inline src/compiletest/runtest.rs]: Given this code block:

```rust
fn main() {
  let i = 1; //~ NOTE: some note
  let j = 2; //~ HELP: some help 
}
```

The current implementation checks for `note` and `help` in one pass through. I'm not sure how it'd be possible using `any` unless we were expecting **both** `help` and `note` on the same line. I agree it's not clean, and I'm willing to clean it up, but would rather do that in a different PR.

File: src/test/compile-fail/issue-25386.rs (hunk 2 above)
Comment (nikomatsakis) [inline src/test/compile-fail/issue-25386.rs]: there really isn't a good way, actually. In this case though I'd just remove the NOTEs. the issue in question (#25386) was an ICE, not a test of diagnostics (I'd not include NOTEs unless the purpose is to test that they are emitted, personally)

File: src/test/compile-fail/lint-unconditional-recursion.rs [hunk 4]
```diff
@@ -9,6 +9,20 @@
...
+//~| NOTE lint level defined here
+//~| NOTE lint level defined here
+//~| NOTE lint level defined here
+//~| NOTE lint level defined here
+//~| NOTE lint level defined here
+//~| NOTE lint level defined here
+//~| NOTE lint level defined here
+//~| NOTE lint level defined here
+//~| NOTE lint level defined here
+//~| NOTE lint level defined here
```
Comment (nikomatsakis) [inline src/test/compile-fail/lint-unconditional-recursion.rs]: oh my, that's a lot of "lint level defined here" messages. but not the fault of this PR I suppose.

**Comment by nikomatsakis** [comment]
left some nits, but seems good

**Comment by nikomatsakis** [comment]
let me know when they're fixed :)

File: src/test/compile-fail/issue-25386.rs (hunk 2 above)
Comment (frewsxcv) [inline src/test/compile-fail/issue-25386.rs]: Cool, that's what I figured, I'll remove them now.

Comment (frewsxcv) [inline src/test/compile-fail/issue-25386.rs]: Removed in the latest force push.

**Comment by frewsxcv** [comment]
> let me know when they're fixed :)

Assuming your only nit was [this one](https://github.com/rust-lang/rust/pull/32263#discussion_r56404772), then it's fixed

File: src/compiletest/runtest.rs (hunk 3 above)
Comment (nikomatsakis) [inline src/compiletest/runtest.rs]: Do it in two passes?

Comment (frewsxcv) [inline src/compiletest/runtest.rs]: Spoke with you on IRC. Going to do it in a follow-up PR.

**Comment by frewsxcv** [comment]
Merge conflicts addressed.
